salt/config/__init__.py

```diff
@@ -3360,13 +3422,18 @@ def get_id(opts, cache_minion_id=False):
         log.debug('Guessing ID. The id can be explicitly set in {0}'
                   .format(os.path.join(salt.syspaths.CONFIG_DIR, 'minion')))
 
-    newid = salt.utils.network.generate_minion_id()
+    if opts.get('id_function'):
+        newid = eval_id_function(opts)
+    else:
+        newid = salt.utils.network.generate_minion_id()
 
     if opts.get('minion_id_lowercase'):
         newid = newid.lower()
         log.debug('Changed minion id {0} to lowercase.'.format(newid))
     if '__role' in opts and opts.get('__role') == 'minion':
-        log.debug('Found minion id from generate_minion_id(): {0}'.format(newid))
+        log.debug('Found minion id from {0}(): {1}'.format(
+            opts.get('id_function') and 'eval_id_function' or
+            'generate_minion_id', newid))
```

Comment:
The `id_function` name is lost here, while is important. I'd include it as well.